crates/wasm-mutate/src/mutators/codemotion/ir/parse_context.rs

```diff
@@ -0,0 +1,249 @@
+        match self.stack.pop() {
+            Some(new_state) => {
+                self.current_parsing = new_state.clone();
+                Ok(new_state)
+            }
+            None => Err(crate::Error::InvalidAstOperation(
+                "Parsing should not be empty".to_string(),
+            )),
+        }
+    }
+
+    /// Push a node to the current parsing
+    ///
+    pub fn push_node_to_current_parsing(&mut self, node: Node) -> usize {
+        let id = self.nodes.len();
+        self.nodes.push(node.clone());
+        self.current_parsing.push(id);
+
+        match node {
+            Node::IfElse {
+                consequent: _,
+                alternative: _,
+                ty: _,
+            } => self.ifs.push(id),
+            Node::Loop(_, _) => self.loops.push(id),
+            Node::Block(_, _) => self.blocks.push(id),
+            _ => {}
+        }
+        id
+    }
+
+    /// Push a new frame,
+    /// * `state` - `If`, `Else`, `Block`, or `Loop` frame type
+    /// * `ty` - Returning type of the frame
+    ///
+    pub fn push_frame(&mut self, state: State, ty: Option<TypeOrFuncType>) {
+        self.frames.push((state, ty))
+    }
+
+    /// Pop frame from the current parsing
+    ///
+    pub fn pop_frame(&mut self) -> crate::Result<(State, Option<TypeOrFuncType>)> {
+        match self.frames.pop() {
+            Some(e) => Ok(e),
+            None => Err(crate::Error::InvalidAstOperation(
+                "Missing parent frame".to_string(),
+            )),
+        }
+    }
+
+    /// Returns the nodes indexes already parsed in the current state
+    ///
+    pub fn get_current_parsing(&self) -> Vec<usize> {
+        self.current_parsing.clone()
+    }
+
+    /// Returns all parsed nodes since the creation of this `ParseContext`
+    /// inetance.
+    ///
+    pub fn get_nodes(&self) -> Vec<Node> {
+        self.nodes.clone()
+    }
+
+    /// Checks if the corrent code parsing has at least one instruction
+    ///
+    pub fn current_code_is_empty(&self) -> bool {
+        self.current_code_range.start == self.current_code_range.end
+    }
+
+    /// Pushes the current code parsing as a `Node::Code` instance
+    ///
+    pub fn push_current_code_as_node(&mut self) -> usize {
+        self.push_node_to_current_parsing(Node::Code {
+            range: self.current_code_range,
+        })
+    }
+
+    /// Resets the current code parsing
+    ///
+    pub fn reset_code_range_at(&mut self, idx: usize) {
+        self.current_code_range = Range::new(idx, idx);
+    }
+
+    /// Augmnents current code parsing to include the next instruction
+    /// in the Wasm code
+    ///
+    pub fn append_instruction_to_current_code(&mut self) {
+        self.current_code_range.end += 1;
+    }
+
+    /// Closes the parsing, creates a `Node::Root` node where the children are the current
+    /// parsed nodes in the current state.
+    ///
+    /// Returns a new Ast instance
+    ///
+    pub fn finish(&mut self) -> Ast {
```

Comment:
This can take ownership of `self` and then we can avoid the clones:

```rust
pub fn finish(self) -> Ast {
    // ...
    Ast {
        // ...
        nodes: self.nodes,
        ifs: self.ifs,
        loops: self.loops,
        blocks: self.blocks,
    }
}
```

I think this would also make the `get_nodes` method obsolete, and we could remove that.